stomp_plugins/src/noise_generators/cartesian_constraints_sampling.cpp

```diff
@@ -0,0 +1,249 @@
+  Eigen::MatrixXd covariance = MatrixXd::Identity(num_timesteps,num_timesteps);
+  covariance = A.transpose() * A;
+  covariance = covariance.fullPivLu().inverse();
+  double max_val = covariance.array().abs().matrix().maxCoeff();
+  covariance /= max_val;
+
+  // preallocating noise data
+  raw_noise_ = Eigen::VectorXd::Zero(CARTESIAN_DOF_SIZE);
+
+  error_code.val = error_code.SUCCESS;
+
+  return true;
+}
+
+bool CartesianConstraintsSampling::setupRobotState(const planning_scene::PlanningSceneConstPtr& planning_scene,
+                                               const moveit_msgs::MotionPlanRequest &req,
+                                               const stomp_core::StompConfiguration &config,
+                                               moveit_msgs::MoveItErrorCodes& error_code)
+{
+  using namespace moveit::core;
+  using namespace utils::kinematics;
+
+  // robot state
+  const JointModelGroup* joint_group = robot_model_->getJointModelGroup(group_);
+  tool_link_ = joint_group->getLinkModelNames().back();
+  state_.reset(new RobotState(robot_model_));
+  robotStateMsgToRobotState(req.start_state,*state_);
+
+  // update kinematic model
+  ik_solver_->setKinematicState(*state_);
+
+  return true;
+}
+
+bool CartesianConstraintsSampling::generateNoise(const Eigen::MatrixXd& parameters,
+                                     std::size_t start_timestep,
+                                     std::size_t num_timesteps,
+                                     int iteration_number,
+                                     int rollout_number,
+                                     Eigen::MatrixXd& parameters_noise,
+                                     Eigen::MatrixXd& noise)
+{
+  using namespace Eigen;
+  using namespace stomp_moveit::utils;
+  using namespace moveit::core;
+
+  if(parameters.rows() != stddev_.size())
+  {
+    ROS_ERROR("%s Number of rows in parameters %i differs from expected number of joints",getName().c_str(),int(parameters.rows()));
+    return false;
+  }
+
+  for(auto t = 0u; t < parameters.cols();t++)
+  {
+    const auto noisy_tool_pose = applyCartesianNoise(parameters.col(t));
+    Eigen::VectorXd result = Eigen::VectorXd::Zero(parameters.rows());
+    if(!ik_solver_->solve(parameters.col(t),noisy_tool_pose,result,tool_goal_tolerance_))
+    {
+      ROS_DEBUG("%s could not compute ik, returning noiseless goal pose",getName().c_str());
+      parameters_noise.col(t) = parameters.col(t);
+    }
+    else
+    {
+      parameters_noise.col(t) = result;
+    }
+  }
+
+  // generating noise
+  noise = parameters_noise - parameters;
+  return true;
+}
+
+Eigen::Affine3d CartesianConstraintsSampling::applyCartesianNoise(const Eigen::VectorXd& reference_joint_pose)
+{
+  using namespace Eigen;
+  using namespace moveit::core;
+  using namespace stomp_moveit::utils;
+
+  const JointModelGroup* joint_group = robot_model_->getJointModelGroup(group_);
+
+  for(auto d = 0u; d < raw_noise_.size(); d++)
```

Comment:
Does the `raw_noise_` realyu need to be a class member? You are only using it here.

Reply:
Yes, it is only use here. In original noise generator is also class member and I think that there is only one disadvantage that method generatedNoise() is called multiple time in each iteration and each rollout and the raw_noise_ must allocated.

Reply:
std::array and local variable